oio/cli/admin/service_check.py

```diff
@@ -0,0 +1,231 @@
+
+    def load_catalog_from_file(self, path):
+        with open(path, "r") as fin:
+            for line in fin:
+                line = line.strip()
+                if not line or line.startswith('#'):
+                    continue
+                try:
+                    t, i, p = line.split()
+                    yield str(t), str(i), int(p), 0
+                except Exception as ex:
+                    self.logger.exception("Failed to loadthe NS services: %s",
+                                          ex)
+
+    @property
+    def logger(self):
+        return self.app.client_manager.logger
+
+    def zookeeper(self):
+        if self._zkcnxstr:
+            return self._zkcnxstr
+
+        conf = self.app.client_manager.sds_conf
+        self._zkcnxstr = conf.get('zookeeper.%s' % self.SRV,
+                                  conf.get('zookeeper'))
+        return self._zkcnxstr
+
+
+class Meta0Check(BaseCheckCommand):
+    SRV = 'meta0'
+
+    def take_action(self, parsed_args):
+        super(Meta0Check, self).take_action(parsed_args)
+        self.load_catalog(parsed_args)
+        import zookeeper
+        from oio.zk.client import get_meta0_paths, get_connected_handles
+        self.logger.debug("Checking the META0 services")
+
+        # TODO: tcp touch to the meta0 services
+
+        # check they are registered in the ZK
+        for zh in get_connected_handles(self.zookeeper()):
+            for p in get_meta0_paths(zh, self.app.options.ns):
+                try:
+                    registered = set()
+                    for n in zookeeper.get_children(zh.get(), p):
+                        v, m = zookeeper.get(zh.get(), p + '/' + n)
+                        registered.add(v)
+                    known = set()
+                    for t, i, p, s in self.filter_services(self.catalog,
+                                                           'meta0'):
+                        known.add('%s:%d' % (i, p))
+                    self.logger.info("meta0 known=%d zk_registered=%d",
+                                     len(known), len(registered))
+                    assert registered == known
+                except Exception as ex:
+                    self.logger.exception(
+                            "Failed to list the M0 from the ZK: %s", ex)
+                finally:
+                    zh.close()
+
+        return ('Status',), [('Ok', )]
+
+
+class Meta1Check(BaseCheckCommand):
+    SRV = 'meta1'
+
+    def take_action(self, parsed_args):
+        super(Meta1Check, self).take_action(parsed_args)
+        self.load_catalog(parsed_args)
+
+        # All the services must have been declared
+        c0 = list(self.filter_services(self.catalog, self.SRV))
+        l0 = list(self.filter_services(self.live, self.SRV))
+        assert len(c0) == len(l0)
+        self.logger.info("All the META1 are alive")
+
+        # They also need a positive score
+        for _, _, _, m1_score in l0:
+            assert m1_score > 0
```

Comment:
Isn't that a bit aggressive? I think Kamel has implemented a facility to change the command's return code and still display results.

Reply:
It's for the `testing` version ;)